invisible_cities/cities/esmeralda.py

```diff
@@ -164,96 +138,102 @@ class HitEnergy(AutoNameEnumBase):
-        #change dtype of columns to match type of variables
-        df = df.apply(lambda x : x.astype(types_dict[x.name]))
 
-        return df, track_hitc
+        if len(hitc.hits)>0:
+            voxels     = plf.voxelize_hits(hitc.hits, vox_size, strict_vox_size, evm.HitEnergy.Ec)
+            mod_voxels = plf.drop_end_point_voxels(voxels, energy_threshold, min_voxels)
+            tracks     = plf.make_track_graphs(mod_voxels)
+
+            vox_size_x = voxels[0].size[0]
+            vox_size_y = voxels[0].size[1]
+            vox_size_z = voxels[0].size[2]
+            del(voxels)
+            #sort tracks in energy
+            tracks     = sorted(tracks, key=plf.get_track_energy, reverse=True)
+
+            track_hits = []
+            for c, t in enumerate(tracks, 0):
+                tID = c
+                energy = plf.get_track_energy(t)
+                length = plf.length(t)
+                numb_of_hits   = len([h for vox in t.nodes() for h in vox.hits])
+                numb_of_voxels = len(t.nodes())
+                numb_of_tracks = len(tracks   )
+
+                pos   = [h.pos for v in t.nodes() for h in v.hits]
+                min_x = min(pos, key = lambda x : x[0])[0]
+                max_x = max(pos, key = lambda x : x[0])[0]
+                min_y = min(pos, key = lambda x : x[1])[1]
+                max_y = max(pos, key = lambda x : x[1])[1]
+                min_z = min(pos, key = lambda x : x[2])[2]
+                max_z = max(pos, key = lambda x : x[2])[2]
+
+                r     = list(map(lambda x: (x[0]**2 + x[1]**2)**0.5, pos))
+                max_r = max(r)
+                min_r = min(r)
+
+                e     = [h.Ec for v in t.nodes() for h in v.hits]
+                ave_pos = np.average(pos, weights=e, axis=0)
+                ave_r   = np.average(r  , weights=e, axis=0)
+                extr1, extr2 = plf.find_extrema(t)
+                extr1_pos = extr1.XYZ
+                extr2_pos = extr2.XYZ
+
+                blob_pos1, blob_pos2 = plf.blob_centres(t, blob_radius)
+
+                e_blob1, e_blob2, hits_blob1, hits_blob2 = plf.blob_energies_and_hits(t, blob_radius)
+                overlap = float(sum(h.Ec for h in set(hits_blob1).intersection(set(hits_blob2))))
+                list_of_vars = [hitc.event, tID, energy, length, numb_of_voxels,
+                                numb_of_hits, numb_of_tracks,
+                                min_x, min_y, min_z, min_r, max_x, max_y, max_z, max_r,
+                                *ave_pos, ave_r, *extr1_pos,
+                                *extr2_pos, *blob_pos1, *blob_pos2,
+                                e_blob1, e_blob2, overlap,
+                                vox_size_x, vox_size_y, vox_size_z]
+
+                df.loc[c] = list_of_vars
+                try:
+                    types_dict
+                except NameError:
+                    types_dict = dict(zip(df.columns, [type(x) for x in list_of_vars]))
```

Comment:
it bothers me a little bit that this thing that is actually constant is being defined here in the middle of everything. But I am not sure the alternative is better. I would try to define the df columns in the top of the file together with this type mapping. BTW, why is it needed?

Reply:
@mmkekic digit: "if i dont change datatype it will be  floats and we dont want that"